### PR TITLE
feat(core): run indexer in worker

### PR DESCRIPTION
**Pull request by EYHN**

### CODE REVIEW

**Comment by EYHN** [comment]
* **#7295** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7295?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#7296** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7296?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7214** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7214?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#7267](https://github.com/toeverything/AFFiNE/pull/7267) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7267?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#7213** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7213?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7212** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7212?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7166** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7166?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @EYHN and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/common/infra/src/modules/workspace/global-schema.ts [hunk 1]
```diff
@@ -3,4 +3,7 @@ import { Schema } from '@blocksuite/store';
 
 export const globalBlockSuiteSchema = new Schema();
 
-globalBlockSuiteSchema.register(AffineSchemas);
+if (globalThis.window !== undefined) {
+  // fixme: let AffineSchemas run without window
```
Comment (forehalo) [inline packages/common/infra/src/modules/workspace/global-schema.ts]: @author

File: packages/frontend/web/src/index.tsx [hunk 2]
```diff
@@ -74,3 +74,7 @@ try {
 } catch (err) {
   console.error('Failed to bootstrap app', err);
 }
+
+window['affine-worker'] = new Worker(new URL('./worker.ts', import.meta.url), {
```
Comment (forehalo) [inline packages/frontend/web/src/index.tsx]: need a wrapper for worker thread, electron may need a different implementation. @pengx17 check it

File: packages/frontend/web/src/worker.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+type Message = {
+  type: 'indexer-crawling';
+  payload: {
+    workspace: WorkspaceMetadata;
+  };
+};
+
+window.onmessage = (event: MessageEvent<Message>) => {
+  if (event.data.type === 'indexer-crawling') {
+    const framework = new Framework();
```
Comment (forehalo) [inline packages/frontend/web/src/worker.ts]: whether do we need a `configureWorker`

File: packages/frontend/web/src/index.tsx (hunk 2 above)
Comment (pengx17) [reply packages/frontend/web/src/index.tsx]: +1. I think we can wrap the indexer messages into RPCs and expose them into existing `apis` handlers for ease of indexing in electron later.

File: packages/frontend/core/src/modules/docs-search/entities/docs-indexer.ts [hunk 4]
```diff
@@ -83,380 +85,214 @@ export class DocsIndexer extends Entity {
...
-      let docTitle = '';
-
-      const blockDocuments: Document<typeof blockIndexSchema>[] = [];
-
-      for (const block of blocks.values()) {
-        const flavour = block.get('sys:flavour')?.toString();
-        const blockId = block.get('sys:id')?.toString();
+      workerOutput = await worker.run({
+        type: 'doc',
+        docBuffer,
```
Comment (forehalo) [inline packages/frontend/core/src/modules/docs-search/entities/docs-indexer.ts:150]: is worker available to access indexdb directly?

File: packages/frontend/core/src/modules/docs-search/worker/in-worker.ts [hunk 5]
```diff
@@ -0,0 +1,308 @@
+import type { AffineTextAttributes } from '@blocksuite/blocks';
+import type { DeltaInsert } from '@blocksuite/inline';
+import { Document } from '@toeverything/infra';
+import { difference } from 'lodash-es';
+import { Array as YArray, Map as YMap, type Text as YText } from 'yjs';
+import { applyUpdate, Doc as YDoc } from 'yjs';
```
Comment (forehalo) [inline packages/frontend/core/src/modules/docs-search/worker/in-worker.ts]: merge two yjs imports

File: packages/frontend/core/src/modules/docs-search/worker/out-worker.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
+import { DebugLogger } from '@affine/debug';
```
Comment (forehalo) [inline packages/frontend/core/src/modules/docs-search/worker/out-worker.ts:1]: what are the meaning of `in/out-worker`?

File: packages/frontend/core/src/modules/docs-search/entities/docs-indexer.ts (hunk 4 above)
Comment (EYHN) [reply packages/frontend/core/src/modules/docs-search/entities/docs-indexer.ts:150]: The storage implementation is injected by DI, probably SQLite

File: packages/frontend/core/src/modules/docs-search/worker/out-worker.ts (hunk 6 above)
Comment (EYHN) [reply packages/frontend/core/src/modules/docs-search/worker/out-worker.ts:1]: in-worker: run in the worker
out-worker: run outside of the worker

File: packages/frontend/core/src/modules/docs-search/entities/docs-indexer.ts (hunk 4 above)
Comment (forehalo) [reply packages/frontend/core/src/modules/docs-search/entities/docs-indexer.ts:150]: if we put the whole job & indexing in worker, we can lowest the data transfer cost. the doc buffer can reaches MB sizes

**Review by forehalo — APPROVED** [review]

**Review by pengx17 — APPROVED** [review]

**Pull request merged**